gui/checkbox.3
.\"	$Csoft: checkbox.3,v 1.17 2005/01/05 04:44:05 vedge Exp $
.\"
.\" Copyright (c) 2002, 2003, 2004, 2005 CubeSoft Communications, Inc.
.\" <http://www.csoft.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd August 20, 2002
.Dt CHECKBOX 3
.Os
.ds vT Agar API Reference
.ds oS Agar 1.0
.Sh NAME
.Nm checkbox
.Nd agar checkbox widget
.Sh SYNOPSIS
.Bd -literal
#include <agar/gui/checkbox.h>
.Ed
.Sh DESCRIPTION
The
.Nm
widget displays a string of text next to a button used to trigger an
.Ft int
binding.
.Sh INHERITANCE HIERARCHY
.Pp
.Xr AG_Object 3 ->
.Xr widget 3 ->
.Nm .
.Sh INITIALIZATION
.nr nS 1
.Ft "AG_Checkbox *"
.Fn AG_CheckboxNew "void *parent" "const char *caption_fmt" "..."
.Pp
.Ft "void"
.Fn AG_CheckboxInit "AG_Checkbox *checkbox" "char *caption"
.Pp
.Ft "void"
.Fn AG_CheckboxToggle "AG_Checkbox *checkbox"
.nr nS 0
.Pp
The
.Fn AG_CheckboxNew
function allocates, initializes, and attaches a
.Nm
widget and accepts a
.Xr printf 3
style format string to set the text of the label.
.Fn AG_CheckboxInit
initializes a
.Nm
widget with the label set to
.Fa caption .
If the length of the label string exceeds
.Dv AG_CHECKBOX_CAPTION_MAX
- 1, the label is truncated.
.Pp
The
.Fn AG_CheckboxToggle
function inverts the current state of
.Fa checkbox .
.Sh BINDINGS
The
.Nm
widget provides the following bindings:
.Pp
.Bl -tag -compact -width "int *value"
.It Va int *state
State (1=enabled, 0=disabled).
.El
.Sh EVENTS
The
.Nm
widget reacts to the following events:
.Pp
.Bl -tag -compact -width 25n
.It window-mousebuttondown
Trigger the state if button is 1, otherwise only give focus to the widget.
.It window-keydown
Trigger the state if key is
.Dv SDLK_RETURN
or
.Dv SDLK_SPACE .
.El
.Pp
The
.Nm
widget generates the following events:
.Pp
.Bl -tag -width 2n
.It Fn checkbox-changed "int state"
Checkbox state changed (1=enabled, 0=disabled).
The
.Sq state
binding remains locked during the event handler's execution.
.El
.Sh SEE ALSO
.Xr agar 3 ,
.Xr button 3 ,
.Xr radio 3 ,
.Xr widget 3 ,
.Xr window 3
.Sh HISTORY
The
.Nm
widget first appeared in Agar 1.0.


trunk/gui/checkbox.c
/*	$Csoft: checkbox.c,v 1.55 2005/10/01 14:15:38 vedge Exp $	*/

/*
 * Copyright (c) 2002, 2003, 2004, 2005 CubeSoft Communications, Inc.
 * <http://www.csoft.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <core/core.h>
#include <core/view.h>

#include "checkbox.h"

#include <gui/window.h>
#include <gui/primitive.h>
#include <gui/label.h>

#include <stdarg.h>
#include <string.h>
#include <errno.h>

static AG_WidgetOps agCheckboxOps = {
	{
		NULL,			/* init */
		NULL,			/* reinit */
		AG_WidgetDestroy,
		NULL,			/* load */
		NULL,			/* save */
		NULL			/* edit */
	},
	AG_CheckboxDraw,
	AG_CheckboxScale
};

#define XSPACING 6

static void checkbox_mousebutton(AG_Event *);
static void checkbox_keydown(AG_Event *);

AG_Checkbox *
AG_CheckboxNew(void *parent, const char *fmt, ...)
{
	char caption[AG_LABEL_MAX];
	AG_Checkbox *cb;
	va_list ap;

	va_start(ap, fmt);
	vsnprintf(caption, sizeof(caption), fmt, ap);
	va_end(ap);

	cb = Malloc(sizeof(AG_Checkbox), M_OBJECT);
	AG_CheckboxInit(cb, caption);
	AG_ObjectAttach(parent, cb);
	return (cb);
}

void
AG_CheckboxInit(AG_Checkbox *cbox, char *caption)
{
	AG_WidgetInit(cbox, "checkbox", &agCheckboxOps, AG_WIDGET_FOCUSABLE);
	AG_WidgetBind(cbox, "state", AG_WIDGET_BOOL, &cbox->state);

	cbox->state = 0;
	cbox->label_su = AG_TextRender(NULL, -1, AG_COLOR(CHECKBOX_TXT_COLOR),
	    caption);
	cbox->label_id = AG_WidgetMapSurface(cbox, cbox->label_su);
	
	AG_SetEvent(cbox, "window-mousebuttondown", checkbox_mousebutton, NULL);
	AG_SetEvent(cbox, "window-keydown", checkbox_keydown, NULL);
}

void
AG_CheckboxDraw(void *p)
{
	AG_Checkbox *cbox = p;

	agPrim.box(cbox,
	    0, 0,
	    AGWIDGET(cbox)->h, AGWIDGET(cbox)->h,
	    AG_WidgetBool(cbox, "state") ? -1 : 1,
	    AG_COLOR(CHECKBOX_COLOR));

	AG_WidgetBlitSurface(cbox, cbox->label_id, AGWIDGET(cbox)->h+XSPACING, 0);
}

static void
checkbox_mousebutton(AG_Event *event)
{
	AG_Checkbox *cbox = AG_SELF();

	AG_WidgetFocus(cbox);
	if (AG_INT(1) == SDL_BUTTON(1))
		AG_CheckboxToggle(cbox);
}

static void
checkbox_keydown(AG_Event *event)
{
	AG_Checkbox *cbox = AG_SELF();

	switch (AG_SDLKEY(1)) {
	case SDLK_RETURN:
	case SDLK_SPACE:
		AG_CheckboxToggle(cbox);
		break;
	default:
		break;
	}
}

void
AG_CheckboxScale(void *p, int rw, int rh)
{
	AG_Checkbox *cb = p;

	if (rh == -1)
		AGWIDGET(cb)->h = cb->label_su->h;
	if (rw == -1)
		AGWIDGET(cb)->w = AGWIDGET(cb)->h + XSPACING + cb->label_su->w;
}

/* Toggle the checkbox state. */
void
AG_CheckboxToggle(AG_Checkbox *cbox)
{
	AG_WidgetBinding *stateb;
	int *state;

	stateb = AG_WidgetGetBinding(cbox, "state", &state);
	*state = !(*state);
	AG_PostEvent(NULL, cbox, "checkbox-changed", "%i", *state);
	AG_WidgetBindingChanged(stateb);
	AG_WidgetUnlockBinding(stateb);
}

